src/Compilers/VisualBasic/Portable/Language/LanguageFeature.vb

```diff
@@ -0,0 +1,293 @@
+        ''' <param name="feature">Language feature to check is available.</param>
+        ''' <param name="options">The parse options being used.</param>
+        ''' <returns>True if the feature's language version is compatible with the specified language version.</returns>
+        <Extension>
+        Private Function IsInLanguageVersion(feature As Feature, options As VisualBasicParseOptions) As Boolean
+            Return CheckVersionNumbers(required:=feature.GetLanguageVersion(), current:=options.LanguageVersion)
+        End Function
+
+        <Extension>
+        Private Function IsInLanguageVersion(feature As Feature, options As VisualBasicCompilation) As Boolean
+            Return CheckVersionNumbers(required:=feature.GetLanguageVersion(), current:=options.LanguageVersion)
+        End Function
+
+        Private Function CheckVersionNumbers(required As LanguageVersion, current As LanguageVersion) As Boolean
+            Return current >= required
+        End Function
+#End Region
+
+#Region "(feature).IsAvailable"
+        ''' <summary>
+        ''' Check to see if a language <paramref name="feature"/> is available with these <paramref name="options"/>.
+        ''' </summary>
+        ''' <remarks>
+        ''' Feature maybe enabled explicitly with a Feature Flag or enabled implicitly via a Language Version.
+        ''' </remarks>
+        ''' <param name="feature">Language feature to check is available.</param>
+        ''' <param name="options">The parse options being used.</param>
+        <Extension>
+        Friend Function IsAvailable(feature As Feature, options As VisualBasicParseOptions) As Boolean
+            Dim flag As String = Nothing
+            If TryGetFeatureFlag(feature, flag) Then Return CheckFeatures(flag, options)
+            Return feature.IsInLanguageVersion(options)
+        End Function
+
+        ''' <summary>
+        ''' Check to see if a language <paramref name="feature"/> is available with this <paramref name="compilation"/>.
+        ''' </summary>
+        ''' <param name="feature">Language feature to check is available.</param>
+        ''' <param name="compilation">The <see cref="VisualBasicCompilation"/> being used.</param>
+        <Extension>
+        Friend Function IsAvailable(feature As Feature, compilation As VisualBasicCompilation) As Boolean
+            Return feature.IsInLanguageVersion(compilation)
+        End Function
+#End Region
+
+#Region "Checking against Options.Features"
+        Private Function TryGetFeatureFlag(f As Feature, ByRef flag As String) As Boolean
+            flag = f.GetFeatureFlag()
+            Return flag IsNot Nothing
+        End Function
+
+        Private Function CheckFeatures(featureFlag As String, options As VisualBasicParseOptions) As Boolean
+            Debug.Assert(featureFlag IsNot Nothing, NameOf(featureFlag) & " can not be nothing.")
+            Return options.Features.ContainsKey(featureFlag)
+        End Function
+#End Region
```

Comment:
Please inline these two methods in the one caller since the inlined versions do not add complexity.

Reply:
How would that affect the method `IsAvailable` being in-line by the JIT?

Reply:
I'm not sure what the difference might be from the JIT, but the extra methods seems unnecessary. Please inline.

Reply:
It should help since the methods are small as well as the methods are in a calls tail-call position. so should reuse the same stack frame. *Not technically the same frame, the stack depth should stay the same.*